### ISSUE TITLE
Hourglass Control -- ULSPH

### ISSUE THREAD

**Issue by Shuaihao-Zhang** [Xiangyu-Hu/SPHinXsys#577]
For elastic case, it seems that we can not use Riemann solver for density. The result is not momentum conservative if we use Riemann solver for density. Below shows the result by using different combinations of Riemann solver for density and pressure.
(1) `Integration1stHalf ` and `Integration2ndHalfDissipativeRiemann`
By using this, we can get correct results.
![1stNoRiemann_2ndDissipativeRiemann](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/e52b83df-f091-406e-863c-0f9f3ee3ec54)
The density and pressure distribution are also smooth.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/7af25d68-2c8d-4d8c-a89b-578a7de0af00)
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/5f05f25e-db97-43d6-bade-7886ea5f9b05)
(2) `Integration1stHalfRiemann ` and `Integration2ndHalfDissipativeRiemann`
The result is not momentum conservative.
![1stRiemann_2ndDissipativeRiemann](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/604209d6-61ac-4471-9d87-7808cd2bc01e)
(2) `Integration1stHalfRiemann ` and `Integration2ndHalfRiemann`
The result is not momentum conservative.
![1stRiemann_2ndRiemann](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/b304d92a-f368-439c-84bc-aee36547e3f0)

For plastic case, using Riemann solver for the density can make the result more stable.
For the Taylor bar case, if we use:
(1) `Integration1stHalf ` and `Integration2ndHalfDissipativeRiemann`
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/798ebc20-2cb1-45ed-812c-408452805423)
We can see that, those particles located at the edges of the column is not uniform.

But if we use:
(2) `Integration1stHalfRiemann ` and `Integration2ndHalfDissipativeRiemann`
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/aaf9d826-39bc-463f-81c5-0badfe160a2d)
We can see those particles are uniform.

For the pressure, we can only use dissipative Riemann, otherwise, if we use:
(3) `Integration1stHalfRiemann ` and `Integration2ndHalfRiemann`
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/58da02a3-62aa-48f5-bdeb-576c786bd012)
The simulation has crashed.

**Comment by Xiangyu-Hu** [comment]
Please also check possible reason for these results.

**Comment by Shuaihao-Zhang** [comment]
I checked the values of parameters when with and without Riemann solver for density.
I found that, when with Riemann solver for density, the density and pressure distribution are more uniform, and the deviation between the maximum and minimum values is smaller, as shown below:
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/dd6351a8-91f8-456f-962d-d4b9150a4701)
Also, when with Riemann solver for density, the force caused by pressure is larger, and the shear force is smaller, as shown below:
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/13e10607-857b-46c1-985f-8b56e0e217e3)
I think the shear force is to oppose the motion. The shear force for the simulation with `Integration1stHalfRiemann` is smaller, so the beam can not stop and go back as the simulation with `Integration1stHalfNoRiemann`, as shown below:
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/0e12b673-83a8-4178-9678-ccd2d41b9a75)
But I haven't found the reason why the shear force is smaller when using `Integration1stHalfRiemann`.
